db/crud.go

```diff
@@ -842,6 +842,204 @@ func (db *Database) PutExistingRev(docid string, body Body, docHistory []string,
 	return doc, err
 }
 
+func (db *Database) validateExistingDoc(doc *Document, importAllowed, docExists bool) error {
+	if !importAllowed && docExists && !doc.HasValidSyncData(db.writeSequences()) {
+		return base.HTTPErrorf(409, "Not imported")
+	}
+	return nil
+}
+
+func (db *Database) validateNewBody(body Body) error {
+	// Reject a body that contains the "_removed" property, this means that the user
+	// is trying to update a document they do not have read access to.
+	if body["_removed"] != nil {
+		return base.HTTPErrorf(http.StatusNotFound, "Document revision is not accessible")
+	}
+
+	// Reject bodies containing user special properties for compatibility with CouchDB
+	if containsUserSpecialProperties(body) {
+		return base.HTTPErrorf(400, "user defined top level properties beginning with '_' are not allowed in document body")
+	}
+	return nil
+}
+
+func (doc *Document) updateWinningRevAndSetDocFlags() (prevCurrentRev string) {
+	var branched, inConflict bool
+	doc.CurrentRev, branched, inConflict = doc.History.winningRevision()
+	doc.setFlag(channels.Deleted, doc.History[doc.CurrentRev].Deleted)
+	doc.setFlag(channels.Conflict, inConflict)
+	doc.setFlag(channels.Branched, branched)
+	if doc.hasFlag(channels.Deleted) {
+		doc.SyncData.TombstonedAt = time.Now().Unix()
+	} else {
+		doc.SyncData.TombstonedAt = 0
+	}
+	return prevCurrentRev
```

Comment:
Looks like this value has been moved up into the calling function. Should remove return value if obsolete.

Reply:
Yup. Missed this, thanks!